File: do-ocr.py
#!/usr/bin/env python3
from pathlib import Path
import subprocess
import json
import sys
import threading
from concurrent.futures import ThreadPoolExecutor

def ocr_file(image):
    bucket_key = image.stem.replace("snap_", '')
    
    try:
        # check if the file name is already in the dictionary, and skip it if so
        with lock:
            if bucket_key in ocr_dict:
                return
        
        # run the ocr command on the file, and capture the output from stdout
        proc = subprocess.run(["/usr/local/bin/OCR", "fr", "false", "true", image.absolute()],
                              stdout=subprocess.PIPE,
                              stderr=subprocess.PIPE)
        
        recognized_text = proc.stdout.decode()
        err = proc.stderr.decode()

        if err:
            print("Erreur", err)
        else:
            with lock:
                print(bucket_key, recognized_text)
                ocr_dict[bucket_key] = recognized_text
                
    except Exception as e:  # catch any exception that might occur during execution
        print("Une erreur est apparu: ", str(e))
        
if __name__ == '__main__':
    lock = threading.Lock()
    
    folder_name = sys.argv[1]
    results_file = sys.argv[2]

    # load the existing dictionary from json file, or create an empty one
    res_file = Path(results_file)
    if res_file.exists():
        with open(res_file, 'r', encoding='utf-8') as f:
            ocr_dict = json.load(f)
    else:
        ocr_dict = {}
    
    # Tweak the threadpool size according to your system's resources
    with ThreadPoolExecutor(max_workers=5) as executor:  # for a 8GB RAM machine, limiting workers to 5 should be fine
        img_path: Path
        for img_path in Path(folder_name).glob("*.png"):
            executor.submit(ocr_file, img_path)
    
    # Write the results back into json file after all threads are done
    with open(results_file, 'w', encoding='utf-8') as f:
        json.dump(ocr_dict, f, ensure_ascii=False, indent=1)

File: doOCR.swift
import Foundation
import PathKit

// Define a dictionary to hold the OCR results
var ocrDict: [String: String] = [:]
let lock = NSLock()

// Function to perform OCR on an image file
func ocrFile(_ image: Path) {
    let bucketKey = image.stem.replacingOccurrences(of: "snap_", with: "")
    
    // Check if the file name is already in the dictionary, and skip it if so
    lock.lock()
    defer { lock.unlock() }
    if ocrDict[bucketKey] != nil { return }
    
    do {
        // Run the OCR command on the file, and capture the output from stdout
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/usr/local/bin/OCR")
        process.arguments = ["fr", "false", "true", image.absolute().string]
        
        let pipe = Pipe()
        process.standardOutput = pipe
        process.standardError = pipe
        
        try process.run()
        process.waitUntilExit()
        
        let data = pipe.fileHandleForReading.readDataToEndOfFile()
        let output = String(data: data, encoding: .utf8) ?? ""
        let errorOutput = String(process.standardError?.readDataToEndOfFile().string ?? "")
        
        if !errorOutput.isEmpty {
            print("Erreur", errorOutput)
        } else {
            lock.lock()
            defer { lock.unlock() }
            print("\(bucketKey) \(output)")
            ocrDict[bucketKey] = output
        }
    } catch {
        print("Une erreur est apparue: ", error.localizedDescription)
    }
}

// Main function to execute the script
if CommandLine.arguments.count != 3 {
    print("Usage: do-ocr.swift <folder_name> <results_file>")
    exit(1)
}

let folderName = CommandLine.arguments[1]
let resultsFile = CommandLine.arguments[2]

// Load the existing dictionary from JSON file, or create an empty one
let resFile = Path(resultsFile)
if resFile.exists {
    do {
        let data = try Data(contentsOf: resFile.url)
        ocrDict = try JSONSerialization.jsonObject(with: data, options: []) as? [String: String] ?? [:]
    } catch {
        print("Failed to load dictionary from file: \(error)")
    }
}

// Get the list of image files in the specified folder
let imgPaths = Path(folderName).glob("*.png").map { $0 }

// Use DispatchQueue for concurrency
DispatchQueue.concurrentPerform(iterations: imgPaths.count) { index in
    ocrFile(imgPaths[index])
}

// Write the results back into JSON file after all tasks are done
do {
    let data = try JSONSerialization.data(withJSONObject: ocrDict, options: .prettyPrinted)
    try data.write(to: resFile.url)
} catch {
    print("Failed to write dictionary to file: \(error)")
}